---
title: 'R exercise: TAD visualization'
author: "Replace with your name"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### 1. Package installation and loading

For this exercise we will use the "HiCBricks" R package (Pal et al. 2020, Bioinformatics), a toolset for Topologically Associated Domain (TAD) data analysis and visualization. First, we install and load the R packages:

```{r}

#BiocManager::install("HiCBricks")

library(HiCBricks)

```

### 2. Data loading

We will work with a dataset included in the "HiCBricks" package. The data corresponds to intra-chromosomal contact maps for Drosophila melanogaster chromosome 3r (Sexton et al. 2012, Cell).

First we have to create a "bricks" object using the information about the structure of the genome:

```{r}

Bintable_path <- system.file(file.path("extdata",
                                       "Bintable_100kb.bins"), package = "HiCBricks")
out_dir <- file.path(tempdir(), "HiCBricks_vignette_test")
dir.create(out_dir)

Create_many_Bricks(BinTable = Bintable_path, 
                   bin_delim=" ", output_directory = out_dir, 
                   file_prefix = "HiCBricks_vignette_test", remove_existing=TRUE, 
                   experiment_name = "HiCBricks vignette test", resolution = 100000)

BrickContainer_dir <- file.path(tempdir(), "HiCBricks_vignette_test")
My_BrickContainer <- load_BrickContainer(project_dir = BrickContainer_dir)
Example_dataset_dir <- system.file("extdata", package = "HiCBricks")

```

Then, we load the Hi-C contact matrix:

```{r}
Chromosomes <- c("chr2L", "chr3L", "chr3R", "chrX")
for (chr in Chromosomes) {
  Matrix_file <- file.path(Example_dataset_dir,
                           paste(paste("Sexton2012_yaffetanay_CisTrans_100000_corrected", 
                                       chr, sep = "_"), "txt.gz", sep = "."))
  Brick_load_matrix(Brick = My_BrickContainer,
                    chr1 = chr,
                    chr2 = chr,
                    resolution = 100000,
                    matrix_file = Matrix_file,
                    delim = " ",
                    remove_prior = TRUE)
}
```

### 3. Call for TADs

The HiCBricks package uses the 2D contact matrices to call TADs:

```{r}
Chromosome <- c("chr2L", "chr3L", "chr3R", "chrX")
di_window <- 10
lookup_window <- 30
TAD_ranges <- Brick_local_score_differentiator(Brick = My_BrickContainer,
                                               chrs = Chromosome,
                                               resolution = 100000,
                                               di_window = di_window,
                                               lookup_window = lookup_window,
                                               strict = TRUE,
                                               fill_gaps = TRUE,
                                               chunk_size = 500)

Name <- paste("LSD",
              di_window,
              lookup_window, sep = "_")
Brick_add_ranges(Brick = My_BrickContainer,
                 ranges = TAD_ranges,
                 rangekey = Name,
                 resolution = 100000)
Brick_list_rangekeys(Brick = My_BrickContainer, resolution = 100000)

TAD_ranges <- Brick_get_ranges(Brick = My_BrickContainer, rangekey = Name,
                               resolution = 100000)

Failsafe_log10 <- function(x){
  x[is.na(x) | is.nan(x) | is.infinite(x)] <- 0
  return(log10(x+1))
}

```

### 4. Visualize the TAD data

There are different ways in which we can explore the TAD data. The plot below shows the HiC data for chromosome 3R between positions 1-10Mbp, stronger red colors represent higher contact scores between different genomic regions.

```{r}

Brick_vizart_plot_heatmap(File = file.path(tempdir(),
                                           "chr3R-1-10MB-normal-colours-log10-valuecap-99.pdf"),
                          Bricks = list(My_BrickContainer),
                          x_coords = "chr3R:1:10000000",
                          y_coords = "chr3R:1:10000000",
                          resolution = 100000,
                          FUN = Failsafe_log10,
                          value_cap = 0.99,
                          legend_title = "Log10 Hi-C signal",
                          palette = "Reds",
                          width = 10,
                          height = 11,
                          return_object = TRUE)

```

In this plot, we can add the TAD boundaries:

```{r}

Brick_vizart_plot_heatmap(File = file.path(tempdir(),
                                           "chr3R-1-10MB-normal-colours-log10-rotate-2-tads.pdf"),
                          Bricks = list(My_BrickContainer),
                          tad_ranges = TAD_ranges,
                          x_coords = "chr3R:1:10000000",
                          y_coords = "chr3R:1:10000000",
                          resolution = 100000,
                          colours = "#230C0F",
                          FUN = Failsafe_log10,
                          value_cap = 0.99,
                          legend_title = "Log10 Hi-C signal",
                          palette = "Reds",
                          width = 10,
                          height = 11,
                          return_object = TRUE)
```

A different way of visualizing:

```{r}

Brick_vizart_plot_heatmap(File = file.path(tempdir(),
                                           "chr3R-1-10MB-normal-colours-log10-rotate.pdf"),
                          Bricks = list(My_BrickContainer),
                          x_coords = "chr3R:1:10000000",
                          y_coords = "chr3R:1:10000000",
                          resolution = 100000,
                          FUN = Failsafe_log10,
                          value_cap = 0.99,
                          distance = 60,
                          legend_title = "Log10 Hi-C signal",
                          palette = "Reds",
                          width = 10,
                          height = 11,
                          rotate = TRUE,
                          return_object = TRUE)

Brick_vizart_plot_heatmap(File = file.path(tempdir(),
                                           "chr3R-1-10MB-normal-colours-log10-rotate-3-tads.pdf"),
                          Bricks = list(My_BrickContainer),
                          tad_ranges = TAD_ranges,
                          x_coords = "chr3R:1:10000000",
                          y_coords = "chr3R:1:10000000",
                          resolution = 100000,
                          colours = "#230C0F",
                          FUN = Failsafe_log10,
                          value_cap = 0.99,
                          distance = 60,
                          legend_title = "Log10 Hi-C signal",
                          palette = "Reds",
                          width = 15,
                          height = 5,
                          line_width = 0.8,
                          cut_corners = TRUE,
                          rotate = TRUE,
                          return_object=TRUE)
```

We can explore the genomic regions associated with our TADs. For example, let's take a look of the TAD located between 2.9-4.2Mb at the NCBI genomic browser (https://www.ncbi.nlm.nih.gov/genome/gdv/browser/genome/?id=GCF_000001215.4).
